File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        n = len(spells)
        m = len(potions)
        potions.sort()
        ans = [0] * n
        for i in range(n):
            need = (success + spells[i] - 1)// spells[i]
            left = 0
            right = m - 1
            while left <= right:
                mid = left + (right - left) // 2
                if potions[mid] < need:
                    left = mid + 1
                else:
                    right = mid - 1
            ans[i] = m - left
        return ans
        

File: 2300-successful-pairs-of-spells-and-potions/2300-successful-pairs-of-spells-and-potions.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        int n = spells.size();
        int m = potions.size();
        vector<int>ans(n);
        sort(potions.begin(), potions.end());
        for (int i = 0; i<n; i++){
            long long temp = (success + spells[i] - 1) / spells[i];
            auto it = lower_bound(potions.begin(), potions.end(), temp);
            ans[i] = potions.end() - it;
        }
        return ans;
    }
};